core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1125,24 +1130,31 @@ class DAGScheduler(
       event.taskInfo.attemptNumber, // this is a task attempt number
       event.reason)
 
-    // The success case is dealt with separately below, since we need to compute accumulator
-    // updates before posting.
-    if (event.reason != Success) {
-      val attemptId = task.stageAttemptId
-      listenerBus.post(SparkListenerTaskEnd(stageId, attemptId, taskType, event.reason,
-        event.taskInfo, event.taskMetrics))
-    }
+    // Reconstruct task metrics. Note: this may be null if the task has failed.
+    val taskMetrics: TaskMetrics =
+      if (event.accumUpdates.nonEmpty) {
+        try {
+          TaskMetrics.fromAccumulatorUpdates(event.accumUpdates)
+        } catch {
+          case NonFatal(e) =>
+            logError(s"Error when attempting to reconstruct metrics for task $taskId", e)
+            null
+        }
+      } else {
+        null
+      }
 
     if (!stageIdToStage.contains(task.stageId)) {
       // Skip all the actions if the stage has been cancelled.
       return
     }
 
+    listenerBus.post(SparkListenerTaskEnd(
```

Comment:
Actually, I guess the old logic was kinda convoluted in the sense that it would report _failures_ of cancelled stages' tasks to the listener bus, but not completions of successes. The new behavior of skipping the listener event for _all_ tasks from cancelled stages, not just successful ones, makes more sense.